domain/application/charm/store/store.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package store
+
+import (
+	"context"
+	"encoding/base32"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/juju/juju/core/objectstore"
+	"github.com/juju/juju/internal/errors"
+	"github.com/juju/juju/internal/uuid"
+)
+
+const (
+	// ErrNotFound is returned when the file is not found.
+	ErrNotFound = errors.ConstError("file not found")
+)
+
+// ObjectStore provides an API for storing objects.
+type ObjectStore interface {
+	// PutAndCheckHash stores data from reader at path, namespaced to the model.
+	// It also ensures the stored data has the correct hash.
+	PutAndCheckHash(ctx context.Context, path string, r io.Reader, size int64, hash string) (objectstore.UUID, error)
+}
+
+// CharmStore provides an API for storing charms.
+type CharmStore struct {
+	objectStore ObjectStore
```

Comment:
You may find some difficulty here.

Presumably, the intention is to construct this when constructing the service and attach the CharmStore and an interface?

However, domainServices had an ObjectStoreGetter, which gets an object store but needs a context, which will not be available during construction.

So here you will need to attach the ObjectStoreGetter and get the object store itself inside your `.Store()` implementation